### PR TITLE
Fix/make posts belong to recipient

### PR DESCRIPTION
**Pull request by Hives**
**MERGE THE OTHER PULL REQUEST (remove '/posts' route) BEFORE THIS ONE!**
I merged that one into this one already, so it makes more sense to
review the other one first.

---

In the posts model we can say a post `belongs_to: recipient`, after
which we can call `post.recipient` to get the recipient user. This allows
us to refactor in various places where we were getting the recipient out
of the database manually.

We can also set up the `post` model to pass `.recipient_id` and
`.recipient_username` through to the `recipient`, which lets us refactor a
couple of other places for readability.

Plus we can remove a couple of conditionals in the post index view now
that you can only get there by specifying a particular user's wall,
since we removed the '/posts' route.

A couple of other minor things too.

### CODE REVIEW

**Comment by Hives** [comment]
[Various refactorings](https://trello.com/c/7HykooYe/80-various-refactorings)

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def create
 
     if user.valid?
       user.save
-      session[:user_id] = user.id
+      log_in(user)
       flash[:success] = "New account created"
```
Comment (j-dav) [inline app/controllers/users_controller.rb:19]: nice use of a helper method

File: app/models/post.rb [hunk 2]
```diff
@@ -1,10 +1,13 @@
 # frozen_string_literal: true
 
 class Post < ApplicationRecord
-  belongs_to :author, :class_name => "User"
-  delegate :email, :to => :author
-  delegate :username, :to => :author
-  
+  belongs_to :author, class_name: "User"
+  delegate :username, to: :author, prefix: :author
+
```
Comment (j-dav) [inline app/models/post.rb:6]: have you written a wiki about this delegate and prefixing?

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -19,7 +19,7 @@
         <ul>
           <% if session[:user_id] %>
             <li>
-              <%= link_to "View posts", posts_path %>
+              <%= link_to "My wall", "/#{current_user.username}" %>
             </li>
```
Comment (j-dav) [inline app/views/layouts/application.html.erb:23]: YAY

File: app/views/posts/index.html.erb [hunk 4]
```diff
@@ -1,23 +1,16 @@
...
+<h1><%= @user.username.capitalize %>'s Wall</h1>
 
 <div class="new-post-button">
-  <% if @user %>
-    <%= link_to new_post_path(:recipient_id => @user.id), class: "btn btn-primary" do %>
-      Create new post
-    <% end %>
+  <%= link_to new_post_path(recipient_id: @user.id), class: "btn btn-primary" do %>
+    Create new post
   <% end %>
```
Comment (j-dav) [inline app/views/posts/index.html.erb:6]: didn't break it this time?

File: app/views/posts/index.html.erb [hunk 5]
```diff
@@ -1,23 +1,16 @@
...
 </div>
 
 <% @posts.each do |post| %>
-
   <div class='post-card'>
     <div class="post-header">
       <div class="post-author">
-        <%= post.username %>
+        <%= post.author_username %>
       </div>
```
Comment (j-dav) [inline app/views/posts/index.html.erb:14]: so this is the prefixing? delegate username to author but call it using author_ prefix?

File: app/models/post.rb (hunk 2 above)
Comment (Hives) [reply app/models/post.rb:6]: no, but good idea

**Review by JRRS1982 — APPROVED** [review]
Looks good, all tests passing and coverage ok.

**Pull request merged**